From 47a6f2eae83290b2199cefcd2ceb77b6f7ddb7b7 Mon Sep 17 00:00:00 2001
From: yuanhao <yuanhao@xiaomi.com>
Date: Mon, 13 Nov 2017 20:24:49 +0800
Subject: [PATCH 034/296] RemoteViews: Fix "Could not copy bitmap to parcel
 blob"

Limit the size of mBitmapCache's mBitmaps, to prevent
opening too many fds when transfering between processes by Parcel.
---
 core/java/android/widget/RemoteViews.java | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/core/java/android/widget/RemoteViews.java b/core/java/android/widget/RemoteViews.java
index 5adbdbe8ab4..11a91438d9f 100644
--- a/core/java/android/widget/RemoteViews.java
+++ b/core/java/android/widget/RemoteViews.java
@@ -54,6 +54,7 @@ import android.os.UserHandle;
 import android.text.TextUtils;
 import android.util.ArrayMap;
 import android.util.Log;
+import android.util.Slog;
 import android.view.LayoutInflater;
 import android.view.LayoutInflater.Filter;
 import android.view.RemotableViewMethod;
@@ -1163,7 +1164,10 @@ public class RemoteViews implements Parcelable, Filter {
     }
 
     private static class BitmapCache {
+        static final int MAX_CACHE_BITMAPS = 50;
+
         ArrayList<Bitmap> mBitmaps;
+        int mCount;
 
         public BitmapCache() {
             mBitmaps = new ArrayList<Bitmap>();
@@ -1174,7 +1178,22 @@ public class RemoteViews implements Parcelable, Filter {
             mBitmaps = new ArrayList<Bitmap>();
             for (int i = 0; i < count; i++) {
                 Bitmap b = Bitmap.CREATOR.createFromParcel(source);
+                addBitmap(b);
+            }
+        }
+
+        private int addBitmap(Bitmap b) {
+            mCount ++;
+            if (mCount > MAX_CACHE_BITMAPS) {
+                int index = (mCount - 1) % MAX_CACHE_BITMAPS;
+                Slog.w(LOG_TAG, "RemoteViews try to cache " + mCount
+                        + " bitmaps, only allows " + MAX_CACHE_BITMAPS
+                        + ", replace bitmap at index " + index);
+                mBitmaps.set(index, b);
+                return index;
+            } else {
                 mBitmaps.add(b);
+                return (mCount - 1);
             }
         }
 
@@ -1185,8 +1204,7 @@ public class RemoteViews implements Parcelable, Filter {
                 if (mBitmaps.contains(b)) {
                     return mBitmaps.indexOf(b);
                 } else {
-                    mBitmaps.add(b);
-                    return (mBitmaps.size() - 1);
+                    return addBitmap(b);
                 }
             }
         }
@@ -1213,7 +1231,7 @@ public class RemoteViews implements Parcelable, Filter {
             for (int i = 0; i < count; i++) {
                 Bitmap b = bitmapsToBeAdded.get(i);
                 if (!mBitmaps.contains(b)) {
-                    mBitmaps.add(b);
+                    addBitmap(b);
                 }
             }
         }
-- 
2.11.0

